src/main/java/com/shapesecurity/shift/es2017/parser/GenericParser.java

```diff
@@ -1568,67 +1694,91 @@ protected Expression parseNewExpression() throws JsError {
+        ArrayList<SpreadElementExpression> args = new ArrayList<>();
+        Maybe<SourceLocation> locationFollowingFirstSpread = Maybe.empty();
         while (true) {
             if (this.match(TokenType.RPAREN) || this.eof()) {
-                return ImmutableList.from(result);
+                return Pair.of(ImmutableList.from(args), locationFollowingFirstSpread);
             }
             SpreadElementExpression arg;
             AdditionalStateT startState = this.startNode();
             if (this.eat(TokenType.ELLIPSIS)) {
                 arg = this.finishNode(startState, new SpreadElement(this.parseAssignmentExpression().left().fromJust()));
+                if (locationFollowingFirstSpread.isNothing()) {
+                    args.add(arg);
+                    if (this.match(TokenType.RPAREN)) {
+                        return Pair.of(ImmutableList.from(args), locationFollowingFirstSpread);
+                    }
+                    locationFollowingFirstSpread = Maybe.of(this.getLocation());
+                    this.expect(TokenType.COMMA);
+                    continue;
+                }
             } else {
-                arg = this.parseAssignmentExpression().left().fromJust();
+                Maybe<Expression> assignmentExpression = this.inheritCoverGrammar(this::parseAssignmentExpressionOrTarget).left();
+                if (assignmentExpression.isNothing()) {
+                    throw this.createUnexpected(this.lookahead);
+                }
+                arg = assignmentExpression.fromJust();
             }
-            result.add(arg);
+            args.add(arg);
             if (!this.eat(TokenType.COMMA)) {
                 break;
             }
         }
-        return ImmutableList.from(result);
+        return Pair.of(ImmutableList.from(args), locationFollowingFirstSpread);
     }
 
     @Nonnull
-    protected Either3<Expression, FormalParameters, AssignmentTarget> parsePrimaryExpression() throws JsError {
+    protected Either3<Expression, Pair<FormalParameters, Boolean>, AssignmentTarget> parsePrimaryExpression() throws JsError {
         if (this.match(TokenType.LPAREN)) {
             return this.parseGroupExpression();
         }
         AdditionalStateT startState = this.startNode();
 
+        if (this.eat(TokenType.ASYNC)) {
+            if (!this.hasLineTerminatorBeforeNext && this.match(TokenType.FUNCTION)) {
+                this.isBindingElement = this.isAssignmentTarget = false;
+                return Either3.left(this.finishNode(startState, this.parseFunctionExpression(false, true)));
+            }
+            return Either3.left(this.finishNode(startState, new IdentifierExpression("async")));
+        }
+
+        if (this.matchIdentifier()) {
+            return Either3.left(this.finishNode(startState, new IdentifierExpression(this.parseIdentifier())));
+        }
+
         switch (this.lookahead.type) {
-            case YIELD:
-                if (this.allowYieldExpression) {
```

Comment:
Was this condition ever hit previously? If so, it looks like it's removal would be a problem.

Reply:
These three cases are handled in `matchIdentifier` a few lines up. The check is still done (just deduplicated).

Reply:
Sorry, `parseIdentifier` calls unexpected identifier, rather than the default case in the switch. matchIdentifier does not throw or check the case, however, this is intentional to allow throwing for invalid yield/await placement.